src/groups.py

```diff
@@ -0,0 +1,380 @@
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        config_data = self._read_file()
+        for key in config_data:
+            data = next(iter(config_data.values()))
+            relation_data = copy.deepcopy(data)
+            self._synchronize(key, data, event)
+            self._add_to_relations(key, relation_data)
+
+    def _read_file(self):
+        """Read the user-group-configuration from the charm config.
+
+        Returns:
+            Data read from the configuration file.
+        """
+        config = self.charm.config["user-group-configuration"]
+        data = yaml.safe_load(config)
+        return data
+
+    def _synchronize(self, key, data, event):
+        """Synchronize data with the Ranger API.
+
+        Args:
+            key: relation_id from configuration file.
+            data: Data to synchronize with the Ranger API.
+            event: configuration changed event.
+        """
+        try:
+            self._sync(data["groups"], "group")
+            self._sync(data["users"], "user")
+            self._sync(data["memberships"], "membership")
+            logger.info(f"Synchronized users and groups for {key}")
+        except RangerServiceException:
+            logger.exception(
+                f"A Ranger Service Exception has occurred while attempting to sync {key}."
+            )
+            event.defer()
+
+    @handle_service_error
+    def _sync(self, config, member_type):
+        """Synchronize apply values with the Ranger API.
+
+        Args:
+            config: Values to synchronize.
+            member_type: The type of Ranger member (group, user or membership).
+        """
+        # Get existing values
+        existing = self._get_existing_values(member_type)
+
+        # Get values to apply
+        apply = self._transform_apply_values(config, member_type)
+
+        # Create members
+        to_create = [item for item in apply if item not in existing]
+        for value in to_create:
+            if member_type == "membership":
+                fields = value
+            if member_type in ["user", "group"]:
+                fields = [
+                    member for member in config if member["name"] == value
+                ][0]
+            self._create_members(member_type, fields, value)
+
+        if member_type == "user":
+            return
+
+        # delete groups and memberships
+        to_delete = [item for item in existing if item not in apply]
+        for value in to_delete:
+            self._delete_members(member_type, value)
+
+    @handle_service_error
+    def _get_existing_values(self, member_type):
+        """Retrieve existing members from the Ranger API.
+
+        Args:
+            member_type: The type of Ranger member (group, user or membership).
+
+        Returns:
+            values: Existing members from the Ranger API.
+        """
+        j = self._query_members(member_type)
```

Comment:
`j` is a bad variable name outside of an iteration variable IMO.

Reply:
Agreed. Updated